src/onegov/search/utils.py

```diff
@@ -42,29 +45,34 @@ def searchable_sqlalchemy_models(
     )
 
 
-def filter_non_base_models(
-    models: 'set[type[T]]'
-) -> 'set[type[T]]':
-    """ Remove model classes that are base classes of other models in the set.
-    Args: models: set of model classes to filter
-    Returns: set: Model classes that are not base classes of any other model
-    in the set.
-
+def filter_for_base_models(
+    models: 'Iterable[type[Searchable]]'
+) -> 'set[type[Searchable]]':
     """
-    non_base_models = set()
+    Filter out models that are polymorphic subclasses of other
+    models in order to save on queries.
 
-    for model in models:
-        is_base = False
-        for other_model in models:
-            if (model is not other_model and issubclass(other_model, model)
-                    and model.__tablename__ == other_model.__tablename__):  # type:ignore[attr-defined]
-                is_base = True
-                break
+    """
+    from onegov.search.mixins import Searchable
 
-        if not is_base:
-            non_base_models.add(model)
+    new_models = set()
 
-    return non_base_models
+    for model in models:
+        i = inspect(model)
+        base_classes = {
+            e.base_mapper.class_ for e in
+            i.base_mapper.self_and_descendants
+            if issubclass(e.base_mapper.class_, Searchable)
+            if e.polymorphic_identity is not None
+            if e.base_mapper.class_ != model
+        }
+        if base_classes:
+            for base in base_classes:
+                new_models.add(base)
+        else:
+            new_models.add(model)
```

Comment:
```suggestion
        i = inspect(model)
        new_models.add(i.base_mapper.class_ if i.polymorphic_on else model)
```

Reply:
Oh, my comment did not make it. 
I tried the above code and it comes back to what I mentioned in my inexistent comment(:laughing:) regarding `File` is not properly handled. `File` is not a subclass of `Searchable` but `GeneralFile` is...

```
(venv) reto@reto-pc:~/workspace/onegov-cloud$ onegov-search --select /onegov_town6/ebikon reindex 
Reindexing onegov_town6/ebikon
2024-12-13 15:44:32 - INFO: Indexing schema onegov_town6-ebikon..
Error psql indexing model 'File': 'EventFile' object has no attribute 'es_skip'
Total objects re-indexed: 12614
took 0:01:07.309937
```
That is why I came up with the latest commit

Reply:
That's what the other suggestion is for. It was missing one line, but it's supposed to filter out non-searchable polymorphic identities.

Reply:
It now indexes `File`,`FileSet` and `FormDefinition` which both are not `Searchable`s. On the other hand I miss `GeneralFile`, `DaypassResource, `RoomResource` and `ItemResource`...

```
(venv) reto@reto-pc:~/workspace/onegov-cloud$ onegov-search --select /onegov_town6/ebikon reindex 
Reindexing onegov_town6/ebikon
2024-12-16 08:15:56 - INFO: Indexing schema onegov_town6-ebikon..
Reindexing model 'CourseEvent'..
Reindexing model 'Event'..
Reindexing model 'Ticket'..
Reindexing model 'ExternalLink'..
Reindexing model 'Directory'..
Reindexing model 'DirectoryEntry'..
Reindexing model 'SettlementRun'..
Reindexing model 'User'..
Reindexing model 'CourseAttendee'..
Reindexing model 'FormDefinition'..
Reindexing model 'Course'..
Reindexing model 'Activity'..
Reindexing model 'Resource'..
Reindexing model 'AgendaItem'..
Reindexing model 'File'..
Reindexing model 'Assembly'..
Reindexing model 'Party'..
Reindexing model 'Newsletter'..
Reindexing model 'ParliamentaryGroup'..
Reindexing model 'Person'..
Reindexing model 'Commission'..
Reindexing model 'Parliamentarian'..
Reindexing model 'Attendee'..
Reindexing model 'AgencyMembership'..
Reindexing model 'Agency'..
Reindexing model 'Page'..
Reindexing model 'Votum'..
Reindexing model 'LegislativePeriod'..
Reindexing model 'FileSet'..
took 0:01:19.406114
```
The filter now adds `base_mapper.class_`es that are not searchable and does not go through all its descendants. For example `class LandsgemeindeFile(File, SearchableFile)` `File` is picked instead of `SearchableFile`. 

I need to revert or skip this optimization of indexing `Page` only instead of `News` and `Topic`

Reply:
No, this is correct and working as intended. When you index, you index all the submodels too and only the ones with a searchable polymorphic identity.

Reply:
The logging message is confusing at this point though, I agree. Maybe switch to `f"Reindexing table  '{mode.__tablename__}'.."`?

Or you generate a log message for each searchable subclass.

Reply:
Yes, the search is not working as the `fts_idx_data` columns are missing for e.g. `Page` model. I added the `fts_idx_data` columns to `Topic` and `News`.

Reply:
> The logging message is confusing at this point though, I agree. Maybe switch to `f"Reindexing table '{model.__tablename__}'.."`?
> 
> Or you generate a log message for each searchable subclass.

Right, I confused myself